pkg/servicemesh/federation/discovery/controller.go

```diff
@@ -116,8 +116,12 @@ func (c *Controller) Run(stopChan <-chan struct{}) {
 	}
 }
 
+func (c *Controller) Start(stopChan <-chan struct{}) {
+	c.Controller.StartController(stopChan, c.HasSynced)
+}
+
 func (c *Controller) HasSynced() bool {
-	return c.Controller.HasSynced()
+	return c.Controller.HasSynced() && c.rm.KubeClient().KubeInformer().Core().V1().ConfigMaps().Informer().HasSynced()
```

Comment:
Can you store the ConfigMap informer as a field in `Controller` and use that field here and in all other places where the federation controller uses this informer? Like it's done here: https://github.com/Maistra/istio/blob/ac3eb3bf6e49b683bdddcb36b7697723c2a96ad7/pilot/pkg/config/kube/ingress/controller.go#L150

This way, it makes it clearer why we're checking the ConfigMap informer here. I think we should use the same pattern for all informers that a particular controller uses.

Reply:
Sure, I will do it.

Reply:
I created a field for ConfigMapInformer, but I didn't refactor other informers in this controller, because they are used just once, and if I stored them in fields, then ResourceManager would no longer make sense, but I don't want to extend the scope of this pull request.

Reply:
I actually thing that `ResourceManager` shouldn't expose `kube.Client`, but each informer individually. Only then can it properly implement `HasSynced()`. 

As you said on Slack, the federation controller also uses Endpoints and ServiceAccounts, but it doesn't ensure that those informers are synced. By exposing each informer individually, it makes it less likely that we'll forget to include an informer in the `HasSynced` logic. But I agree this is a broader change, so it can be done in a separate PR.

Reply:
The more I think about this, the clearer it becomes that we need to refactor the ResourceManager, as I said above. This is an important source of current and future bugs.

Reply:
Ok, so I will expose only those informers that are used instead of exposing kube client. I will also extend HasSynced to check all informers.